pkg/platform/runtime/setup/setup.go

```diff
@@ -351,7 +365,12 @@ func (s *Setup) installFromBuildLog(buildResult *model.BuildResult, artifacts ar
 	}
 	defer conn.Close()
 
-	buildLog, err := buildlog.New(artifacts, conn, s.events, *buildResult.Recipe.RecipeID)
+	alreadyBuilt := make(map[artifact.ArtifactID]struct{})
+	for _, d := range downloads {
+		alreadyBuilt[d.ArtifactID] = struct{}{}
+	}
+	artifactLogManager := buildlog.NewArtifactLogs(ctx, s.events)
+	buildLog, err := buildlog.New(artifacts, alreadyBuilt, conn, artifactLogManager, s.events, *buildResult.Recipe.RecipeID)
```

Comment:
buildlog taking artifactLogManager as an arg looks like a bit of a red flag to me. You seem to ONLY initialize it to then pass it into buildlog.New(), and they're both part of the same internal package. Doesn't seem we need to be exposing `buildlog.NewArtifactLogs` at all?

Reply:
Agreed that it is awkward! See my comment above.  Adding the line 372 to the `New` function creates a problem during testing though, because the `ArtifactLogs` struct creates actual websocket connections.  And for testing purposes I tried to make the implementation of `BuildLog` independent of the connection implementation such that it can be faked easily...

Reply:
What I usually do in cases like this is use a separate `NewButMoreCustom()` function that is used from the regular `New()` call as well as from tests.

Reply:
Here is how that would look like: 5053602a4  I must admit that I am not a big fan of this, as then suddenly the `BuildLog` owns the connections (and has to close them too).  IMO it was cleaner before, even though it results in more code/effort for the caller of this constructor.  But it makes the boundaries of `BuildLog`'s implementation more confined (maybe even more "unixy").